components/prism-psl.js

```diff
@@ -0,0 +1,28 @@
+Prism.languages.psl = {
+	'comment': [
+		{
+			pattern: /#.*/,
+			greedy: true
+		}
+	],
+	'string': [{
+		pattern: /"(?:\\"|[^"\r\n]|[\r\n])*"/,
+		greedy: true,
+		inside: {
+			'symbol': /\\[ntrbA-Z"\\]/
+		}
+	},
+	{
+		pattern: /<<<([a-zA-Z_]\w*)[\r\n](?:.*[\r\n])*?\1\b/,
+		alias: 'heredoc-string',
+		greedy: true
+	}],
+
+	'keyword': /\b(?:export|function|local|exit|last|next|requires|return|until|case|default|do|elsif|else|for|foreach|if|switch|while|line|word|__single|__multi)\b/,
+	'constant': /\b(?:ALARM|CHART_ADD_GRAPH|CHART_DELETE_GRAPH|CHART_DESTROY|CHART_LOAD|CHART_PRINT|EOF|FALSE|False|false|NO|No|no|OK|OFFLINE|PSL_PROF_LOG|R_CHECK_HORIZ|R_CHECK_VERT|R_CLICKER|R_COLUMN|R_FRAME|R_ICON|R_LABEL|R_LABEL_CENTER|R_LIST_MULTIPLE|R_LIST_MULTIPLE_ND|R_LIST_SINGLE|R_LIST_SINGLE_ND|R_MENU|R_POPUP|R_POPUP_SCROLLED|R_RADIO_HORIZ|R_RADIO_VERT|R_ROW|R_SCALE_HORIZ|R_SCALE_VERT|R_SPINNER|R_TEXT_FIELD|R_TEXT_FIELD_LABEL|R_TOGGLE|TRUE|True|true|TRIM_LEADING|TRIM_TRAILING|TRIM_REDUNDANT|TRIM_LEADING_AND_TRAILING|VOID|WARN)\b/,
+	'variable': /\b(?:errno|exit_status|PslDebug)\b/,
+	'function': /\b(?:acos|add_diary|annotate|annotate_get|asctime|asin|atan|atexit|ascii_to_ebcdic|batch_set|blackout|cat|ceil|chan_exists|change_state|close|code_cvt|cond_signal|cond_wait|console_type|convert_base|convert_date|convert_locale_date|cos|cosh|create|destroy_lock|dump_hist|date|destroy|difference|dget_text|dcget_text|ebcdic_to_ascii|encrypt|event_archive|event_catalog_get|event_check|event_query|event_range_manage|event_range_query|event_report|event_schedule|event_trigger|event_trigger2|execute|exists|exp|fabs|floor|fmod|full_discovery|file|fopen|ftell|fseek|grep|get_vars|getenv|get|get_chan_info|get_ranges|get_text|gethostinfo|getpid|getpname|history_get_retention|history|index|int|is_var|intersection|isnumber|internal|in_transition|join|kill|length|lines|lock|lock_info|log|loge|log10|matchline|msg_check|msg_get_format|msg_get_severity|msg_printf|msg_sprintf|ntharg|num_consoles|nthargf|nthline|nthlinef|num_bytes|print|proc_exists|process|popen|printf|pconfig|poplines|pow|PslExecute|PslFunctionCall|PslFunctionExists|PslSetOptions|random|read|readln|refresh_parameters|remote_check|remote_close|remote_event_query|remote_event_trigger|remote_file_send|remote_open|remove|replace|rindex|sec_check_priv|sec_store_get|sec_store_set|set_alarm_ranges|set_locale|share|sin|sinh|sleep|sopen|sqrt|srandom|subset|set|substr|system|sprintf|sort|subset|snmp_agent_config|_snmp_debug|snmp_agent_stop|snmp_agent_start|snmp_h_set|snmp_h_get_next|snmp_h_get|snmp_set|snmp_walk|snmp_get_next|snmp_get|snmp_config|snmp_close|snmp_open|snmp_trap_receive|snmp_trap_ignore|snmp_trap_listen|snmp_trap_send|snmp_trap_raise_std_trap|snmp_trap_register_im|splitline|strcasecmp|str_repeat|trim|tail|tan|tanh|time|tmpnam|tolower|toupper|trace_psl_process|text_domain|unlock|unique|union|unset|va_arg|va_start|write)\b/,
```

Comment:
Is this huge list really necessary? Wouldn't the approach languages like C and JavaScript take be better? Those languages use the function call syntax (`name()`) to detect function definitions and calls.

Not only will this drastically reduce the file size of the language, but it will also support user-defined functions.

This only has 2 problems:

1. This won't detect [`export function`](https://docs.bmc.com/docs/PATROLAgent/2008/export-938228605.html) statements, but that's easy to add (see suggested code).
2. [`foreach`](https://docs.bmc.com/docs/PATROLAgent/2008/foreach-938228607.html). This is difficult. We could either accept these false positives or we could try to match these variables before we match functions. Kinda like this:
	
	```js
	'foreach-variable': {
		pattern: /(\bforeach\s+(?:(?:\w+\b|"(?:\\.|[^\\"])*")\s+){0,2})[_a-zA-Z]\w*(?=\s*\()/,
		lookbehind: true,
		greedy: true
	},
	```
	I'm undecided as to what would be better.

```suggestion
	'function': [
		/\b[_a-z]\w*(?=\s*\()/i,
		{
			pattern: /(\bfunction\s+)[_a-zA-Z]\w*\b/,
			lookbehind: true
		}
	],
```

Reply:
I really want to highlight these builtin functions, as they are different in nature from user-defined functions. These builtin functions are not part of a library made of user-defined functions, they are really part of the language definition itself.

I'm really in favor of highlighting these specifically. That's how it's currently done in VSCode, so that would be consistent.

Reply:
I see. In that case, let's keep your current solution.

Reply:
This is what it looks like in VSCode with the [psl-lang extension](https://marketplace.visualstudio.com/items?itemName=sentrysoftware.vscode-psl-lang&ssr=false#overview):
![image](https://user-images.githubusercontent.com/32521698/105915827-70143000-6030-11eb-83b3-34f8264b4e76.png)

Reply:
Oh, so the grammar for VSCode highlights them like this. In that case, let's do the same! 

I updated my suggestion. The new suggestion will tokenize all built-in functions as `builtin` and all user-defined functions as `function`. This should closely match VSCode's behaviour.